include/lbann/layers/transform/cross_grid_sum.hpp

```diff
@@ -158,8 +172,43 @@ class cross_grid_sum_layer final : public data_type_layer<TensorDataType>
   void bp_setup_gradient_wrt_inputs(El::Int mini_batch_size) final
   {
     int rank = El::mpi::Rank(this->get_subgrid_comm());
-    const auto& gradient_wrt_output = this->get_prev_error_signals(rank);
-    auto& gradient_wrt_input = this->get_error_signals(rank);
+    auto parents = this->get_parent_layers();
+    auto childs = this->get_child_layers();
+
+    int tag = -1;
+    int tag_parent = -1;
+    int tag_child = -1; 
+    int count =0;
+    for(int i=0; i<parents.size();i++)
+    {
+      if (this->get_error_signals(i).Grid().InGrid())
+        tag_parent = parents[i]->get_grid_tag();
+
+    }
+    tag = tag_parent-1;
+    for(int i=0; i<childs.size();i++)
+    {
+      if (this->get_prev_error_signals(i).Grid().InGrid())
+        tag_child = childs[i]->get_grid_tag();
+    }
+
+    // int tag_parent = parents[rank]->get_grid_tag();
+    // int tag_child = childs[rank]->get_grid_tag(); 
+    
+    const auto& gradient_wrt_output = this->get_prev_error_signals(tag);
+    auto& gradient_wrt_input = this->get_error_signals(tag);
+
+
+    int gradient_wrt_output_Height = gradient_wrt_output.Height();
+    int gradient_wrt_output_Width = gradient_wrt_output.Width();
+    for(int i=0; i<childs.size();i++)
+    {
+      auto* const gradient_wrt_input_cast = dynamic_cast<
```

Comment:
What if this is not the type? Needs error message.

Reply:
I can add it here but there are lot of places where we are just converting like this

Reply:
OK, but I am not sure about the STAR/VC/ELEMENT type being always the same.